### PR TITLE
ARM64-SVE: Add checks for MOVPRFX sequencing (#105514)

### LINKED ISSUE #105514: Arm64/Sve: Validation for movprfx in emitter
As noted in https://github.com/dotnet/runtime/pull/100743#discussion_r1563072598 . Everytime we emit SVE instruction, we should check the last emitted instruction and if it is `movprfx`, then validate the things required for `movprfx` placement.

image

### PR DESCRIPTION
**Pull request by a74nh**
As per the Arm Arm, `MOVPRFX` instructions must be followed by an SVE instruction of specific type, otherwise behaviour is undefined.

Add a function which checks that two instructions in sequence are valid. For now this only cares when the first instruction is a `MOVPRFX`, but the function could be expanded to check for any pairs, should we need to.

Function is called after code generation in a loop which cycles through all `IG`s and all instructions.
This ensures the stream is valid after peephole type optimisations.

Alternatively, this function could be called from `emitInsSanityCheck()`. `EMIT_BACKWARDS_NAVIGATION` can then be used to get the previous instruction. However, defining this adds an extra variable to `instrDesc`, reducing the small immediate size from 7bits to 2bits.

Alternatively, this function could be inside every `emitIns_` function, allowing `emitLastIns` to be used for the previous instruction. However, that is a lot of call sites and some could easily be missed. In addition a peephole optimisation could still occur after this point.

Currently requires #106125 for all SVE tests to pass.
AddSequentialAcross tests disabled until #106180 is fixed.

Resolves: https://github.com/dotnet/runtime/issues/105514

### CODE REVIEW

**Comment by a74nh** [comment]
@dotnet/arm64-contrib @kunalspathak

File: src/coreclr/jit/emit.cpp
Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:7107]: i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted *last*). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Suggested some changes. Also, can you also confirm all stress tests are passing?

File: src/coreclr/jit/emit.cpp
Comment (a74nh) [reply src/coreclr/jit/emit.cpp:7107]: > this will not catch case where `movprfx` was in previous IG and following SVE instruction is in new IG.

That should be ok. `prevId` is defined outside the IG loop and isn't reset ever. So on the next IG, `prevId` should be the last instruction in the previous `IG`.

> i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted _last_). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

I remember trying to do that when we did the peephole optimisations, and there was push back. I can't remember the issues around it. But maybe that is why `EMIT_BACKWARDS_NAVIGATION` was added instead?

Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:7107]: > prevId is defined outside the IG loop and isn't reset ever

that should be fine then.

> that is why EMIT_BACKWARDS_NAVIGATION was added instead

It was added specifically to track more than 1 last instruction for xarch zero extension peephole IIRC. I realized that we also have `emitRemoveLastInstruction()` so perhaps doing the validation once everything is emitted makes sense.

File: src/coreclr/jit/emitarm64.cpp
Comment (a74nh) [reply src/coreclr/jit/emitarm64.cpp]: > i was hoping to see lot more validation for registers of firstId

That should be all at the bottom of the function. I'll double check that I have everything, because the text you have above is a much more verbose version than what I was using.

Comment (kunalspathak) [reply src/coreclr/jit/emitarm64.cpp]: looks like you have most of the things covered. We need to add validation for:

- MOVPRFX cannot be the last instruction in a sequence
- The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)
- MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18597]:
```suggestion
            assert(!"Got unexpected instruction format after MOVPRFX");
```

File: src/coreclr/jit/emitarm64.cpp
Comment (a74nh) [reply src/coreclr/jit/emitarm64.cpp]: > * MOVPRFX cannot be the last instruction in a sequence

Done.

> * The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)

Looking at all the formats, they are all <Pg>/M or unpredicated. I've added some extra checking to make sure they correspond to the correct movprfx instruction.

> * MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

That's done in the switch. Eg:
`            assert(secondId->idReg1() != secondId->idReg3());`
I had to do it there because you need the format to know what types the registers are.

File: src/coreclr/jit/emit.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emit.cpp:7116]:
```suggestion
            assert(prevId->idIns() != INS_sve_movprfx);
```

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18446]:
```suggestion
       
```

not needed any more if you just check it in the caller.

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18456]:
```suggestion
        // unpredicated version
```

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18460]:
```suggestion
    // Quoted sections are taken from the Arm manual.
```

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18658]: This assert can be outside if-else.

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18655]:
```suggestion
        
```

no need of this validation. we are already doing it above.

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18637]:
```suggestion
    if (movprefxIsPredicated)
```

File: src/coreclr/jit/emitarm64sve.cpp
Comment (kunalspathak) [inline src/coreclr/jit/emitarm64sve.cpp:18647]: can this be true for unpredicated version as well?

**Review by kunalspathak — DISMISSED** [review]
Added some more comments on validation.

File: src/coreclr/jit/emitarm64sve.cpp
Comment (a74nh) [reply src/coreclr/jit/emitarm64sve.cpp:18639]: The unpredicated `movprfx` doesn't care about element sizes, it is just `MOVPRFX <Zd>, <Zn>`.

**Comment by a74nh** [comment]
Remove the changes to `GenerateHWIntrinsicTests_Arm.cs`.

This PR needs to wait for #106292 (and #106125) until it can be merged

**Comment by a74nh** [comment]
> This PR needs to wait for #106292 (and #106125) until it can be merged

Those are merged now. There is nothing blocking this PR.

However - when running the stress testing I see some more errors from this PR. Will investigate.

**Comment by amanasifkhalid** [comment]
> However - when running the stress testing I see some more errors from this PR. Will investigate.

@a74nh I'm assuming you're getting a clean run now, right?

File: src/coreclr/jit/emit.cpp
Comment (amanasifkhalid) [inline src/coreclr/jit/emit.cpp:7100]: Nit:
```suggestion
#endif // defiined(DEBUG) && defined(TARGET_ARM64)
```

File: src/coreclr/jit/emit.cpp
Comment (amanasifkhalid) [inline src/coreclr/jit/emit.cpp:7118]: Nit:
```suggestion
#endif // defined(DEBUG) && defined(TARGET_ARM64)
```

File: src/coreclr/jit/emitarm64sve.cpp
Comment (amanasifkhalid) [inline src/coreclr/jit/emitarm64sve.cpp:18438]: Nit:
```suggestion
 *  Sanity check two instructions are valid when placed next to each other
```

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp
Comment (amanasifkhalid) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2574]: What is this syntax doing? It looks like the result of the ternary operator isn't being used.

**Comment by amanasifkhalid** [comment]
>I'll raise a bug and take a look at it - I have a reasonable idea of what it might be.
This PR shouldn't need to be blocked on it though.

Got it, thanks!

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp
Comment (a74nh) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2574]: That change should have been on the line above - the `HWIntrinsicImmOpHelper` constructor needs to take in the number of instructions that will be generated in each stage of the loop.

**Comment by a74nh** [comment]
> I'll raise a bug and take a look at it - I have a reasonable idea of what it might be.

This was due to changes in #106125. I've fixed in this PR because it's related and simple enough - that PR changed the type of `movprfx` emitted when it shouldn't have. 

Also fixed up review comments.

Just checking now that all stress testing passes now.

**Review by amanasifkhalid — APPROVED** [review]
LGTM if stress tests pass. Thanks!

**Comment by a74nh** [comment]
> LGTM if stress tests pass. Thanks!

Full stress testing for all APIs passing now.
ro: https://gist.github.com/a74nh/b690f44a2589873fce3066592d5141fa
and same for r too.

**Pull request merged**
